diff -r e96c4a40fdf9 src/solaris/native/sun/awt/fontpath.c
--- openjdk/jdk/src/solaris/native/sun/awt/fontpath.c	Wed Aug 23 14:14:49 2017 -0400
+++ openjdk/jdk/src/solaris/native/sun/awt/fontpath.c	Thu Aug 24 12:14:41 2017 -0400
@@ -1239,6 +1239,7 @@
                 && (strcmp((char*)fontformat, "TrueType") != 0)
 #if defined(__linux__) || defined(_AIX)
                 && (strcmp((char*)fontformat, "Type 1") != 0)
+                && (strcmp((char*)fontformat, "CFF") != 0)
 #endif
              ) {
                 continue;
