--- openjdk/jdk/makefiles/CopyFiles.gmk.orig	2013-08-06 12:47:30.000000000 -0400
+++ openjdk/jdk/makefiles/CopyFiles.gmk	2013-08-06 13:19:21.000000000 -0400
@@ -302,7 +302,7 @@
   # The main problem is deciding whether to use aliases for the VMs that are not
   # present and the current position is that we add aliases for client and server, but
   # not for minimal.
-  # To do: should this also support, -zero and -zeroshark?
+  # To do: should this also support -zeroshark?
 
   CLIENT_AND_SERVER := $(and $(findstring true,$(JVM_VARIANT_SERVER)),$(findstring true,$(JVM_VARIANT_CLIENT)))
 
@@ -335,6 +335,11 @@
 	        $(PRINTF) "-minimal KNOWN\n">>$(@)
 	        $(PRINTF) "-server ALIASED_TO -minimal\n">>$(@)
 	        $(PRINTF) "-client ALIASED_TO -minimal\n">>$(@)
+            else
+                ifeq ($(JVM_VARIANT_ZERO),true)
+	            $(PRINTF) "-server KNOWN\n">>$(@)
+	            $(PRINTF) "-client IGNORED\n">>$(@)
+                endif
             endif
         endif
     endif
@@ -346,6 +351,8 @@
 	$(call install-file)
 endif
 
+
+
 COPY_FILES += $(JVMCFG)
 
 ##########################################################################################
