diff -up jdk8/corba/make/common/shared/Defs-java.gmk.ppc openjdk/corba/make/common/shared/Defs-java.gmk
--- jdk8/corba/make/common/shared/Defs-java.gmk.ppc	2012-03-24 11:44:34.450072733 +0100
+++ jdk8/corba/make/common/shared/Defs-java.gmk	2012-03-24 11:44:53.300073137 +0100
@@ -79,9 +79,9 @@ JAVAC_JVM_FLAGS =
 
 # 64-bit builds require a larger thread stack size.
 ifeq ($(ARCH_DATA_MODEL), 32)
-  JAVAC_JVM_FLAGS    += -J-XX:ThreadStackSize=768
+  JAVAC_JVM_FLAGS    += -J-XX:ThreadStackSize=1152
 else
-  JAVAC_JVM_FLAGS    += -J-XX:ThreadStackSize=1536
+  JAVAC_JVM_FLAGS    += -J-XX:ThreadStackSize=1664
 endif
 JAVAC_JVM_FLAGS    += $(JAVA_TOOLS_FLAGS:%=-J%)
 
