diff -up jdk8/common/autoconf/platform.m4.s390 jdk8/common/autoconf/platform.m4
--- jdk8/common/autoconf/platform.m4.s390	2013-04-30 07:30:55.368691627 -0400
+++ jdk8/common/autoconf/platform.m4	2013-04-30 07:31:51.168692356 -0400
@@ -60,6 +60,18 @@ AC_DEFUN([PLATFORM_EXTRACT_VARS_FROM_CPU
       VAR_CPU_BITS=64
       VAR_CPU_ENDIAN=big
        ;;
+    s390)
+      VAR_CPU=s390
+      VAR_CPU_ARCH=s390
+      VAR_CPU_BITS=32
+      VAR_CPU_ENDIAN=big
+       ;;
+    s390x)
+      VAR_CPU=s390x
+      VAR_CPU_ARCH=s390x
+      VAR_CPU_BITS=64
+      VAR_CPU_ENDIAN=big
+       ;;
     sparc)
       VAR_CPU=sparc
       VAR_CPU_ARCH=sparc
