diff --git openjdk.orig/hotspot/make/linux/makefiles/vm.make openjdk.orig/hotspot/make/linux/makefiles/vm.make
--- openjdk.orig/hotspot/make/linux/makefiles/vm.make
+++ openjdk/hotspot/make/linux/makefiles/vm.make
@@ -121,6 +121,7 @@
 CFLAGS += $(CFLAGS/NOEX)
 
 # Extra flags from gnumake's invocation or environment
+ASFLAGS += $(EXTRA_CFLAGS)
 CFLAGS += $(EXTRA_CFLAGS)
 LFLAGS += $(EXTRA_CFLAGS) $(EXTRA_LDFLAGS)
 
