diff --git a/java-1.8.0-openjdk-1.8.0.242.b06-0.0.ea.fc32.x86_64/openjdk/hotspot/src/cpu/aarch64/vm/interp_masm_aarch64.hpp b/java-1.8.0-openjdk-1.8.0.242.b06-0.0.ea.fc32.x86_64/openjdk/hotspot/src/cpu/aarch64/vm/interp_masm_aarch64.hpp
index 22beab0..f3572e1 100644
--- a/java-1.8.0-openjdk-1.8.0.242.b06-0.0.ea.fc32.x86_64/openjdk/hotspot/src/cpu/aarch64/vm/interp_masm_aarch64.hpp
+++ b/java-1.8.0-openjdk-1.8.0.242.b06-0.0.ea.fc32.x86_64/openjdk/hotspot/src/cpu/aarch64/vm/interp_masm_aarch64.hpp
@@ -40,8 +40,6 @@ class InterpreterMacroAssembler: public MacroAssembler {
  protected:
 
  protected:
-  using MacroAssembler::call_VM_leaf_base;
-
   // Interpreter specific version of call_VM_base
   using MacroAssembler::call_VM_leaf_base;
 
