diff -urN jdk8/hotspot/make/bsd/makefiles/gcc.make jdk8/hotspot/make/bsd/makefiles/gcc.make
--- jdk8/hotspot/make/bsd/makefiles/gcc.make	2013-07-26 14:47:20.000000000 +0200
+++ jdk8/hotspot/make/bsd/makefiles/gcc.make	2013-07-29 19:56:54.217369655 +0200
@@ -125,10 +125,10 @@
 
 ARCHFLAG = $(ARCHFLAG/$(BUILDARCH))
 ARCHFLAG/i486    = -m32 -march=i586
-ARCHFLAG/amd64   = -m64
+ARCHFLAG/amd64   =  
 ARCHFLAG/ia64    =
 ARCHFLAG/sparc   = -m32 -mcpu=v9
-ARCHFLAG/sparcv9 = -m64 -mcpu=v9
+ARCHFLAG/sparcv9 =   -mcpu=v9
 ARCHFLAG/zero    = $(ZERO_ARCHFLAG)
 
 # Darwin-specific build flags
diff -urN jdk8/hotspot/make/bsd/makefiles/sparcv9.make jdk8/hotspot/make/bsd/makefiles/sparcv9.make
--- jdk8/hotspot/make/bsd/makefiles/sparcv9.make	2013-07-26 14:47:20.000000000 +0200
+++ jdk8/hotspot/make/bsd/makefiles/sparcv9.make	2013-07-29 19:56:54.230369493 +0200
@@ -21,7 +21,7 @@
 # questions.
 #  
 
-# gcc 4.0 miscompiles this code in -m64
+# gcc 4.0 miscompiles this code in  
 OPT_CFLAGS/macro.o = -O0
 
 CFLAGS += -D_LP64=1
diff -urN jdk8/hotspot/make/bsd/makefiles/sparcWorks.make jdk8/hotspot/make/bsd/makefiles/sparcWorks.make
--- jdk8/hotspot/make/bsd/makefiles/sparcWorks.make	2013-07-26 14:47:20.000000000 +0200
+++ jdk8/hotspot/make/bsd/makefiles/sparcWorks.make	2013-07-29 19:56:54.187370032 +0200
@@ -37,7 +37,7 @@
 
 ARCHFLAG = $(ARCHFLAG/$(BUILDARCH))
 ARCHFLAG/i486    = -m32
-ARCHFLAG/amd64   = -m64
+ARCHFLAG/amd64   =  
 
 CFLAGS     += $(ARCHFLAG)
 AOUT_FLAGS += $(ARCHFLAG)
diff -urN jdk8/hotspot/make/linux/makefiles/gcc.make jdk8/hotspot/make/linux/makefiles/gcc.make
--- jdk8/hotspot/make/linux/makefiles/gcc.make	2013-07-26 14:47:20.000000000 +0200
+++ jdk8/hotspot/make/linux/makefiles/gcc.make	2013-07-29 19:56:54.372367708 +0200
@@ -92,11 +92,11 @@
 
 ARCHFLAG = $(ARCHFLAG/$(BUILDARCH))
 ARCHFLAG/i486    = -m32 -march=i586
-ARCHFLAG/amd64   = -m64
+ARCHFLAG/amd64   =  
 ARCHFLAG/aarch64 =
 ARCHFLAG/ia64    =
 ARCHFLAG/sparc   = -m32 -mcpu=v9
-ARCHFLAG/sparcv9 = -m64 -mcpu=v9
+ARCHFLAG/sparcv9 =   -mcpu=v9
 ARCHFLAG/arm     =  -fsigned-char
 ARCHFLAG/zero    = $(ZERO_ARCHFLAG)
 ifndef E500V2
diff -urN jdk8/hotspot/make/linux/makefiles/sparcv9.make jdk8/hotspot/make/linux/makefiles/sparcv9.make
--- jdk8/hotspot/make/linux/makefiles/sparcv9.make	2013-07-26 14:47:20.000000000 +0200
+++ jdk8/hotspot/make/linux/makefiles/sparcv9.make	2013-07-29 19:56:54.389367494 +0200
@@ -21,7 +21,7 @@
 # questions.
 #  
 
-# gcc 4.0 miscompiles this code in -m64
+# gcc 4.0 miscompiles this code in  
 OPT_CFLAGS/macro.o = -O0
 
 CFLAGS += -D_LP64=1
diff -urN jdk8/hotspot/make/linux/makefiles/sparcWorks.make jdk8/hotspot/make/linux/makefiles/sparcWorks.make
--- jdk8/hotspot/make/linux/makefiles/sparcWorks.make	2013-07-26 14:47:20.000000000 +0200
+++ jdk8/hotspot/make/linux/makefiles/sparcWorks.make	2013-07-29 19:56:54.350367983 +0200
@@ -37,7 +37,7 @@
 
 ARCHFLAG = $(ARCHFLAG/$(BUILDARCH))
 ARCHFLAG/i486    = -m32
-ARCHFLAG/amd64   = -m64
+ARCHFLAG/amd64   =  
 
 CFLAGS     += $(ARCHFLAG)
 AOUT_FLAGS += $(ARCHFLAG)
diff -urN jdk8/hotspot/make/solaris/makefiles/gcc.make jdk8/hotspot/make/solaris/makefiles/gcc.make
--- jdk8/hotspot/make/solaris/makefiles/gcc.make	2013-07-26 14:47:20.000000000 +0200
+++ jdk8/hotspot/make/solaris/makefiles/gcc.make	2013-07-29 19:56:53.855374206 +0200
@@ -79,9 +79,9 @@
 ARCHFLAG = $(ARCHFLAG/$(BUILDARCH))
 
 ARCHFLAG/sparc   = -m32 -mcpu=v9
-ARCHFLAG/sparcv9 = -m64 -mcpu=v9
+ARCHFLAG/sparcv9 =   -mcpu=v9
 ARCHFLAG/i486    = -m32 -march=i586
-ARCHFLAG/amd64   = -m64 -march=k8
+ARCHFLAG/amd64   =   -march=k8
 
 
 # Optional sub-directory in /usr/lib where BUILDARCH libraries are kept.
diff -urN jdk8/hotspot/make/solaris/makefiles/sparcWorks.make jdk8/hotspot/make/solaris/makefiles/sparcWorks.make
--- jdk8/hotspot/make/solaris/makefiles/sparcWorks.make	2013-07-26 14:47:20.000000000 +0200
+++ jdk8/hotspot/make/solaris/makefiles/sparcWorks.make	2013-07-29 19:56:53.833374483 +0200
@@ -109,11 +109,11 @@
 ARCHFLAG_OLD/sparc   = -xarch=v8plus
 ARCHFLAG_NEW/sparc   = -m32 -xarch=sparc
 ARCHFLAG_OLD/sparcv9 = -xarch=v9
-ARCHFLAG_NEW/sparcv9 = -m64 -xarch=sparc
+ARCHFLAG_NEW/sparcv9 =   -xarch=sparc
 ARCHFLAG_OLD/i486    =
 ARCHFLAG_NEW/i486    = -m32
 ARCHFLAG_OLD/amd64   = -xarch=amd64
-ARCHFLAG_NEW/amd64   = -m64
+ARCHFLAG_NEW/amd64   =  
 
 # Select the ARCHFLAGs and other SS12 (5.9) options
 ifeq ($(shell expr $(COMPILER_REV_NUMERIC) \>= 509), 1)
diff -urN jdk8/hotspot/src/share/tools/hsdis/Makefile jdk8/hotspot/src/share/tools/hsdis/Makefile
--- jdk8/hotspot/src/share/tools/hsdis/Makefile	2013-07-26 14:47:20.000000000 +0200
+++ jdk8/hotspot/src/share/tools/hsdis/Makefile	2013-07-29 19:56:46.286469350 +0200
@@ -45,7 +45,7 @@
 endif
 endif
 CFLAGS/sparcv9	+= -xarch=v9
-CFLAGS/amd64	+= -m64
+CFLAGS/amd64	+=  
 CFLAGS		+= $(CFLAGS/$(ARCH))
 DLDFLAGS	+= -G
 LDFLAGS         += -ldl
@@ -77,8 +77,8 @@
 CFLAGS+= -DTARGET_ARCH_aarch64
 endif
 ifdef LP64
-CFLAGS/sparcv9	+= -m64
-CFLAGS/amd64	+= -m64
+CFLAGS/sparcv9	+=  
+CFLAGS/amd64	+=  
 else
 ARCH=$(ARCH1:amd64=i386)
 CFLAGS/i386	+= -m32
@@ -102,8 +102,8 @@
 ARCH1=$(CPU:x86_64=amd64)
 ARCH=$(ARCH1:i686=i386)
 ifdef LP64
-CFLAGS/sparcv9  += -m64
-CFLAGS/amd64    += -m64
+CFLAGS/sparcv9  +=  
+CFLAGS/amd64    +=  
 else
 ARCH=$(ARCH1:amd64=i386)
 CFLAGS/i386     += -m32
diff -urN jdk8/hotspot/test/runtime/6929067/Test6929067.sh jdk8/hotspot/test/runtime/6929067/Test6929067.sh
--- jdk8/hotspot/test/runtime/6929067/Test6929067.sh	2013-07-26 14:47:21.000000000 +0200
+++ jdk8/hotspot/test/runtime/6929067/Test6929067.sh	2013-07-29 19:56:54.452366701 +0200
@@ -43,7 +43,7 @@
 grep "64-Bit" vm_version.out > ${NULL}
 if [ "$?" = "0" ]
 then
-  COMP_FLAG="-m64"
+  COMP_FLAG=" "
 else
   COMP_FLAG="-m32"
 fi
diff -urN jdk8/jdk/make/common/Defs-linux.gmk jdk8/jdk/make/common/Defs-linux.gmk
--- jdk8/jdk/make/common/Defs-linux.gmk	2013-07-26 14:58:06.000000000 +0200
+++ jdk8/jdk/make/common/Defs-linux.gmk	2013-07-29 19:57:12.975133862 +0200
@@ -190,8 +190,8 @@
 CFLAGS_REQUIRED_amd64   += -fno-omit-frame-pointer -D_LITTLE_ENDIAN
 CFLAGS_REQUIRED_i586    += -fno-omit-frame-pointer -D_LITTLE_ENDIAN
 CFLAGS_REQUIRED_ia64    += -fno-omit-frame-pointer -D_LITTLE_ENDIAN
-CFLAGS_REQUIRED_sparcv9 += -m64 -mcpu=v9
-LDFLAGS_COMMON_sparcv9  += -m64 -mcpu=v9
+CFLAGS_REQUIRED_sparcv9 +=   -mcpu=v9
+LDFLAGS_COMMON_sparcv9  +=   -mcpu=v9
 CFLAGS_REQUIRED_sparc   += -m32 -mcpu=v9
 LDFLAGS_COMMON_sparc    += -m32 -mcpu=v9
 CFLAGS_REQUIRED_arm     += -fsigned-char -D_LITTLE_ENDIAN
diff -urN jdk8/jdk/make/common/Defs-macosx.gmk jdk8/jdk/make/common/Defs-macosx.gmk
--- jdk8/jdk/make/common/Defs-macosx.gmk	2013-07-26 14:58:06.000000000 +0200
+++ jdk8/jdk/make/common/Defs-macosx.gmk	2013-07-29 19:57:13.085132479 +0200
@@ -113,13 +113,13 @@
 #    We need this frame pointer to make it easy to walk the stacks.
 #    This should be the default on X86, but ia64, and x86_64
 #    may not have this as the default.
-CFLAGS_REQUIRED_x86_64  += -m64 -fno-omit-frame-pointer -D_LITTLE_ENDIAN
-LDFLAGS_COMMON_x86_64   += -m64
+CFLAGS_REQUIRED_x86_64  +=   -fno-omit-frame-pointer -D_LITTLE_ENDIAN
+LDFLAGS_COMMON_x86_64   +=  
 CFLAGS_REQUIRED_i586    += -m32 -fno-omit-frame-pointer -D_LITTLE_ENDIAN
 LDFLAGS_COMMON_i586     += -m32
-CFLAGS_REQUIRED_ia64    += -m64 -fno-omit-frame-pointer -D_LITTLE_ENDIAN
-CFLAGS_REQUIRED_sparcv9 += -m64 -mcpu=v9
-LDFLAGS_COMMON_sparcv9  += -m64 -mcpu=v9
+CFLAGS_REQUIRED_ia64    +=   -fno-omit-frame-pointer -D_LITTLE_ENDIAN
+CFLAGS_REQUIRED_sparcv9 +=   -mcpu=v9
+LDFLAGS_COMMON_sparcv9  +=   -mcpu=v9
 CFLAGS_REQUIRED_sparc   += -m32 -mcpu=v9
 LDFLAGS_COMMON_sparc    += -m32 -mcpu=v9
 CFLAGS_REQUIRED_arm     += -fsigned-char -D_LITTLE_ENDIAN
diff -urN jdk8/jdk/make/common/shared/Compiler-sun.gmk jdk8/jdk/make/common/shared/Compiler-sun.gmk
--- jdk8/jdk/make/common/shared/Compiler-sun.gmk	2013-07-26 14:58:06.000000000 +0200
+++ jdk8/jdk/make/common/shared/Compiler-sun.gmk	2013-07-29 19:57:13.041133032 +0200
@@ -86,12 +86,12 @@
 XARCH_OPTION_OLD/32 =
 XARCH_OPTION_OLD/64 =
 XARCH_OPTION_NEW/32 = -m32
-XARCH_OPTION_NEW/64 = -m64
+XARCH_OPTION_NEW/64 =  
 # Lint options are slightly different
 LINT_XARCH_OPTION_OLD/32 =
 LINT_XARCH_OPTION_OLD/64 =
 LINT_XARCH_OPTION_NEW/32 = -m32
-LINT_XARCH_OPTION_NEW/64 = -m64
+LINT_XARCH_OPTION_NEW/64 =  
 ifeq ($(ARCH_FAMILY), sparc)
   ifdef VIS_NEEDED
     XARCH_OPTION_OLD/32 += -xarch=v8plusa
diff -urN jdk8/jdk/make/sun/xawt/Makefile jdk8/jdk/make/sun/xawt/Makefile
--- jdk8/jdk/make/sun/xawt/Makefile	2013-07-26 14:58:07.000000000 +0200
+++ jdk8/jdk/make/sun/xawt/Makefile	2013-07-29 19:57:13.978121255 +0200
@@ -233,7 +233,7 @@
 SIZERS_C = $(SIZER_64_C)
 SIZES = $(WRAPPER_GENERATOR_DIR)/sizes.64
 ifdef CROSS_COMPILE_ARCH
-CFLAGS_64 = -m64
+CFLAGS_64 =  
 endif
 endif # 32
 endif # !macosx
