diff -r 57824090cf28 src/share/bin/java.c
--- openjdk/jdk/src/share/bin/java.c	Fri Feb 05 17:41:45 2016 +0000
+++ openjdk/jdk/src/share/bin/java.c	Thu Feb 18 03:43:51 2016 +0000
@@ -699,22 +699,6 @@
     }
 }
 
-/*
- * This is called if the jvmtype returned by CheckJvmType above
- * didn't exist. In this case if it was the default VM and we
- * selected the server vm we will allow it to use the client vm
- * instead.
- */
-char *
-GetAltJvmType(char *jvmtype)
-{
-    if ((knownVMs[0].flag == VM_IF_SERVER_CLASS)) {
-        if (jvmtype == knownVMs[0].server_class+1) return knownVMs[0].name+1;
-        if (jvmtype == knownVMs[0].name+1) return knownVMs[0].server_class+1;
-    }
-    return NULL;
-}
-
 /* copied from HotSpot function "atomll()" */
 static int
 parse_size(const char *s, jlong *result) {
diff -r 57824090cf28 src/share/bin/java.h
--- openjdk/jdk/src/share/bin/java.h	Fri Feb 05 17:41:45 2016 +0000
+++ openjdk/jdk/src/share/bin/java.h	Thu Feb 18 03:43:51 2016 +0000
@@ -164,7 +164,6 @@
  */
 jint ReadKnownVMs(const char *jvmcfg, jboolean speculative);
 char *CheckJvmType(int *argc, char ***argv, jboolean speculative);
-char *GetAltJvmType(char *jvmtype);
 void AddOption(char *str, void *info);
 
 enum ergo_policy {
diff -r 57824090cf28 src/solaris/bin/aarch64/jvm.cfg
--- openjdk/jdk/src/solaris/bin/aarch64/jvm.cfg	Fri Feb 05 17:41:45 2016 +0000
+++ openjdk/jdk/src/solaris/bin/aarch64/jvm.cfg	Thu Feb 18 03:43:51 2016 +0000
@@ -31,6 +31,5 @@
 # "-XXaltjvm=<jvm_dir>" option, but that too is unsupported
 # and may not be available in a future release.
 #
--client IF_SERVER_CLASS -server
 -server KNOWN
--minimal KNOWN
+-client IGNORE
